pkgs/build-support/docker/store-path-to-layer.sh

```diff
@@ -11,39 +11,35 @@ echo "Creating layer #$layerNumber for $@"
 mkdir -p "$layerPath"
 
 # Make sure /nix and /nix/store appear first in the archive.
+#
 # We create the directories here and use them because
 # when there are other things being added to the
 # nix store, tar could fail, saying,
 # "tar: /nix/store: file changed as we read it"
 mkdir -p nix/store
-tar -cf "$layerPath/layer.tar"  \
-    --mtime="@$SOURCE_DATE_EPOCH" \
-    --owner=0 --group=0 \
-    --transform='s,nix,/nix,' \
-    nix
-
-# We change into the /nix/store in order to avoid a similar
-# "file changed as we read it" error as above. Namely,
-# if we use the absolute path of /nix/store/123-pkg
-# and something new is added to the nix store while tar
-# is running, it will detect a change to /nix/store and
-# fail. Instead, if we cd into the nix store and copy
-# the relative nix store path, tar will ignore changes
-# to /nix/store. In order to create the correct structure
-# in the tar file, we transform the relative nix store
-# path to the absolute store path.
-for storePath in "$@"; do
-  n=$(basename "$storePath")
-  tar -C /nix/store -rpf "$layerPath/layer.tar" \
+
+# Then we change into the /nix/store in order to
+# avoid a similar "file changed as we read it" error
+# as above. Namely, if we use the absolute path of
+# /nix/store/123-pkg and something new is added to the nix
+# store while tar is running, it will detect a change to
+# /nix/store and fail. Instead, if we cd into the nix store
+# and copy the relative nix store path, tar will ignore
+# changes to /nix/store. In order to create the correct
+# structure in the tar file, we transform the relative nix
+# store path to the absolute store path.
+tarhash=$(
+  basename -a "$@" |
+    tar -cp nix \
+      -C /nix/store --verbatim-files-from --files-from - \
       --hard-dereference --sort=name \
       --mtime="@$SOURCE_DATE_EPOCH" \
       --owner=0 --group=0 \
-      --transform="s,$n,/nix/store/$n," \
-      $n
-done
-
-# Compute a checksum of the tarball.
-tarhash=$(tarsum < $layerPath/layer.tar)
+      --transform 's,^nix(/|$),/nix/,' \
+      --transform 's,^[^/],/nix/store/\0,rS' |
```

Comment:
hm, it seems my `sed` powers are now powerful enough ;) 
- what is the purpose of `^[^/]`? Why `^` is not sufficient?
- I also don't understand why `rS` flags are added? What is the default behavior, especially regarding the `r` flag?

Reply:
> it seems my sed powers are now powerful enough ;)

I think that's only because they're `tar`'s variant of `sed` :).

> what is the purpose of ^[^/]? Why ^ is not sufficient?

The transforms get applied one after another, and since the first one converts `^nix` to `^/nix`, I had to make sure that the second transform won't match to that. Otherwise that `nix/` directory at the root would become `/nix/store/nix`. 

> I also don't understand why rS flags are added? What is the default behavior, especially regarding the r flag?

The previous code was only replacing the exact store path(`s,$n,/nix/store/$n`), but since we now archive multiple paths at the same time, we need to prepend `/nix/store` to everything. But turns out those transformations are applied to symlink targets too, and even when they are relative (eg. `...libpng.so.1 -> libpng.so` would become `...libpng.so.1 -> /nix/store/libpng.so`). That `S` flag disables transforming the symlinks. However, just setting `S` overrides _all_ of the default flags(`rsh` is the default), so it doesn't apply the transformations to the regular files either. Only setting both `r` and `S` makes it so that the transformation is applied to regular files, but not to the symlinks. 

Here's the doc I used: https://www.gnu.org/software/tar/manual/html_section/tar_51.html

Reply:
wow... ok. Thanks for explanations!